### ISSUE TITLE
GEP: Add support for CORS

### ISSUE THREAD

**Issue by arkodg** [kubernetes-sigs/gateway-api#1767]
**What would you like to be added**:
The ability to configure [CORS ](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) policies on a HTTPRoute

**Why this is needed**:
[CORS ](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) is needed to relax specific restrictions of [SOP](https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy) associated with requests sent from a web browser.

To learn more visit https://enable-cors.org/index.html

How some common data plane implementations that do this today
* Envoy - https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/cors_filter
* NGINX - https://enable-cors.org/server_nginx.html
* Many more in - https://enable-cors.org/server.html

Here is an example of existing high level APIs

* Contour - https://projectcontour.io/docs/v1.21.1/config/cors/
* Istio - https://istio.io/latest/docs/reference/config/networking/virtual-service/#CorsPolicy
* Envoy Gateway - https://gateway.envoyproxy.io/latest/user/cors/
* Ingress NGINX - https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#enable-cors

Relates to https://github.com/envoyproxy/gateway/issues/567

**Comment by youngnick** [comment]
I think that this one could either be done with a Policy or with fields. A Policy design should probably wait for #1565 to merge, but a review of implementations and what CORS configuration they support, and what's common between them (like I did for #1744) should allow us to have a better design here.

**Comment by Xunzhuo** [comment]
Yes, this needs more infros before getting it started, I am going to resolve other assigned high priority issues first.

**Comment by youngnick** [comment]
I think for this one, we need to talk about _how_ to do it first, before we start, which will need probably a small GEP. @Xunzhuo, I'm going to unassign this for now, as it's reasonably large, and you've got a lot on your plate.

**Comment by arkodg** [comment]
related https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#enable-cors

**Comment by shaneutt** [comment]
We've marked this one as `v1.0.0` and placed it on [the road to ga](https://github.com/orgs/kubernetes-sigs/projects/30) as something to do next. For this to actually make it in GA however, we probably need an owner/champion for it to start driving it forward and starting a GEP _fairly soon_. Is anyone up for that task?

/help

**Comment by zhaohuabing** [comment]
Just added CORS to the SecurityPolicy in Envoy Gateway. I can try to come up an initial GEP from what we learned from Envoy Gateway.

**Comment by tao12345666333** [comment]
CORS is a common requirement, as we have implemented in the Kubernetes Ingress-NGINX project. However, I believe that Policy Attachment could be a more flexible and practical option for certain scenarios.

**Comment by lianglli** [comment]
It would be great if HTTPRouteFilter had a field (E.g., HTTPRouteFilterCORS) to set CORS config (i.e., cors_allow_origin, cors_allow_methods, cors_allow_headers, cors_max_age, cors_allow_credentials and cors_expose_headers). 

Then, if a HTTPRoute has the HTTPRouteFilterCORS, then the gateway will :
* return 204 with specific CORS headers to client for HTTP OPTION;
* return 200 with specific CORS headers to client.

**Comment by lianglli** [comment]
> We've marked this one as `v1.0.0` and placed it on [the road to ga](https://github.com/orgs/kubernetes-sigs/projects/30) as something to do next. For this to actually make it in GA however, we probably need an owner/champion for it to start driving it forward and starting a GEP _fairly soon_. Is anyone up for that task?
> 
> /help

@shaneutt 
It is a common feature for HTTP gateway.
If this requires a GEP, I would be like to start working on it.

**Comment by lianglli** [comment]
The CORS were discussed at Kubecon last week. @mikemorris @arkodg  I will start a GEP for CORS that focuses on background and past implementations first.

**Comment by youngnick** [comment]
Earlier on this issue I suggested that it would require work on the _how_ to do this, which is true but misleading.

CORS is a reasonably complex feature that may be different between data paths (that is the proxies that actually pass the traffic).

So the next steps for this GEP (which can proceed at _any time_ and is not affected by the Gateway API release cycle) are:

* Someone requests to have the work assigned to them on this issue (I've removed the current assignee because there have been no updates for some time).
* That person generates a new PR to the `geps/` directory in the repo. This PR must do a few things:
  * copy the template in the `gep-696` directory to a new `gep-1767` directory
  * update all references to GEP-696 to GEP-1767, in both the Markdown and YAML files in there
  * Mark this new GEP as the `Provisional` state, both in the Markdown and the YAML files
  * Fill out the title (`CORS Support`), TLDR, Goals, Non-Goals and Introduction sections of the GEP _only_. Other sections can be left as they are in the template or filled out with "To be completed later" or similar.
  * The most important part of this update is the Introduction. This section should explain what CORS is and why people want to be able to configure it, and, even more importantly, explain what each relevant data plane (proxy) does to configure CORS today. See [GEP-1742](https://gateway-api.sigs.k8s.io/geps/gep-1742/) for a similar example, although this CORS GEP should also include checking if any Gateway API implementations support CORS already, and how they configure it (I believe that Envoy Gateway and Istio already do, but I'm not sure).
  
 The purpose of this initial Provisional update is to ensure that everyone talking about CORS in Gateway API has the same understanding of the current state of the art around configuring and using CORS in both Gateway API implementations and their underlying data plane proxies. This will mean that, at a future date, we can look at doing the further work to push this GEP to Implementable and Experimental (which _will_ be subject to the usual planning cycle and freeze periods). Up until the Provisional state is finished, though, all updates to this GEP document will _not_ be covered by Gateway API change freezes.

Marking as good-first-issue with this todo list.

/good-first-issue

**Comment by lianglli** [comment]
/assign @lianglli

Sorry for the late update. I will create a GEP asap.

**Comment by youngnick** [comment]
/reopen

We shouldn't close GEPs until the feature moves to Standard.
